code/android-identity-library/src/main/java/com/adobe/marketing/mobile/identity/IdentityExtension.java

```diff
@@ -0,0 +1,2070 @@
+import com.adobe.marketing.mobile.services.ServiceProvider;
+import com.adobe.marketing.mobile.util.DataReader;
+import com.adobe.marketing.mobile.util.StringUtils;
+import com.adobe.marketing.mobile.util.TimeUtils;
+import com.adobe.marketing.mobile.util.URLBuilder;
+import com.adobe.marketing.mobile.util.UrlUtils;
+
+import java.net.HttpURLConnection;
+import java.util.*;
+
+/**
+ * IdentityExtension class is responsible for interactions with the ECID Service
+ * <p>
+ * The IdentityExtension handles the following use cases:
+ * <ol>
+ *     <li>Syncing new visitor identifiers to the ECID Service</li>
+ *     <li>Modifying a base URL and returning with ECID Service values to facilitate hybrid app communication</li>
+ *     <li>Returning a list of known identifiers for the user</li>
+ * </ol>
+ * <p>
+ * The IdentityExtension listens for the following {@link Event}s:
+ * <ol>
+ *     <li>{@code EventType.HUB} - {@link EventSource#SHARED_STATE}</li>
+ *     <li>{@link EventType#IDENTITY} - {@link EventSource#REQUEST_IDENTITY}</li>
+ * </ol>
+ * <p>
+ * The IdentityExtension dispatches the following {@code Events}:
+ * <ol>
+ *     <li>{@link EventType#ANALYTICS} - {@link EventSource#REQUEST_CONTENT}</li>
+ *     <li>{@code EventType.IDENTITY} - {@link EventSource#RESPONSE_IDENTITY}</li>
+ *     <li>{@code EventType.CONFIGURATION} - {@link EventSource#REQUEST_CONTENT}</li>
+ * </ol>
+ * <p>
+ */
+final public class IdentityExtension extends Extension {
+    private static final String LOG_SOURCE = "IdentityExtension";
+    private HitQueuing hitQueue;
+    private static boolean pushEnabled = false;
+    private static final Object _pushEnabledMutex = new Object();
+    private ConfigurationSharedStateIdentity latestValidConfig;
+    private final NamedCollection namedCollection;
+
+    private String mid;
+    private String advertisingIdentifier;
+    private String pushIdentifier;
+    private String blob;
+    private String locationHint;
+    private long lastSync;
+    private long ttl;
+    private List<VisitorID> customerIds;
+    private MobilePrivacyStatus privacyStatus = IdentityConstants.Defaults.DEFAULT_MOBILE_PRIVACY;
+    private boolean hasSynced = false;
+
+    /**
+     * Construct the extension and initialize with the {@code ExtensionApi}.
+     *
+     * @param extensionApi the {@link ExtensionApi} this extension will use
+     */
+    IdentityExtension(@NonNull ExtensionApi extensionApi) {
+        super(extensionApi);
+        this.namedCollection = ServiceProvider.getInstance().getDataStoreService().
+                getNamedCollection(DataStoreKeys.IDENTITY_PROPERTIES_DATA_STORE_NAME);
+    }
+
+    @VisibleForTesting
+    IdentityExtension(@NonNull ExtensionApi extensionApi, @Nullable NamedCollection namedCollection, @NonNull HitQueuing hitQueue) {
+        super(extensionApi);
+        this.namedCollection = namedCollection;
+        this.hitQueue = hitQueue;
+    }
```

Comment:
Can you chain these constructor calls. I worry because diverged constructors may lead to initialization duplication and sometimes missed initializations when they are modified in the future.